frontend/awx/resources/templates/WorkflowVisualizer/Topology.tsx

```diff
@@ -0,0 +1,228 @@
+  );
+
+  const createVisualization = useCallback(() => {
+    const newVisualization = new Visualization();
+    newVisualization.setFitToScreenOnLayout(true);
+    newVisualization.registerComponentFactory(baselineComponentFactory);
+    newVisualization.registerLayoutFactory(
+      (type: string, graph: Graph) =>
+        new DagreLayout(graph, {
+          rankdir: 'LR',
+          ranker: 'network-simplex',
+          ranksep: 150,
+        })
+    );
+    newVisualization.fromModel(graphModel, false);
+    newVisualization.addEventListener(SELECTION_EVENT, handleSelectedNode);
+    return newVisualization;
+  }, [baselineComponentFactory, handleSelectedNode]);
+
+  useEffect(() => {
+    const layoutNodes = results.map((node) => {
+      return {
+        ...node,
+        runAfterTasks: [],
+      };
+    });
+
+    return setLayout(() => [...layoutNodes]);
+  }, [results]);
+
+  useEffect(() => {
+    if (!visualizationRef.current) {
+      visualizationRef.current = createVisualization();
+    }
+  }, [createVisualization, visualization]);
+
+  useEffect(() => {
+    if (!visualization) {
+      return;
+    }
+
+    const nodes: GraphNode[] = layout.map((n) => {
+      const processNodes = (nodesArray: number[]) => {
+        nodesArray.forEach((id) => {
+          const afterNode = layout.find((afterNode) => afterNode.id.toString() === id.toString());
+          if (afterNode) {
+            if (afterNode.runAfterTasks) {
+              afterNode.runAfterTasks.push(n.id.toString());
+            } else {
+              afterNode.runAfterTasks = [n.id.toString()];
+            }
+          }
+        });
+      };
+      if (n.success_nodes.length > 0) {
+        processNodes(n.success_nodes);
+      }
+      if (n.failure_nodes.length > 0) {
+        processNodes(n.failure_nodes);
+      }
+      if (n.always_nodes.length > 0) {
+        processNodes(n.always_nodes);
+      }
+
+      const node: GraphNode = {
+        id: n.id.toString(),
+        type: 'node',
+        label: n.summary_fields?.unified_job_template?.name || 'UNDEFINED',
+        width: NODE_DIAMETER,
+        height: NODE_DIAMETER,
+        shape: NodeShape.ellipse,
+        runAfterTasks: n.runAfterTasks,
+        data: {
+          jobType: n.summary_fields?.unified_job_template?.unified_job_type,
+        },
+      };
+
+      return node;
+    });
+
+    const edges = nodes.flatMap((targetNode) => {
+      if (!targetNode.runAfterTasks) return [];
+
+      return targetNode.runAfterTasks.map((sourceNodeId) => {
+        const tagLabel = {
+          [NodeStatus.success]: t('On success'),
+          [NodeStatus.danger]: t('On fail'),
+          [NodeStatus.info]: t('On always'),
+        };
+
+        const sourceNode = layout.find((node) => node.id === Number(sourceNodeId));
+        const targetNodeId = targetNode.id;
+
+        let status = NodeStatus.info;
+        if (sourceNode) {
+          if (sourceNode.success_nodes?.includes(Number(targetNodeId))) {
```

Comment:
Success nodes and failure nodes are array values that come directly from the api so that to me is a reason not to transform it because either way we would would have to loop through each array for succes_nodes, failure_nodes, and always_nodes.

Reply:
Yes, but that would be a one time and not within an inner loop.
We can ignore this for now, but if performance testing shows an issue, this is a spot that can be optimized.